[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Boyu-Zhang-UOI/pml-f2026-notebooks/blob/main/05-deep-learning-pytorch/05_lightning_same_model.ipynb)

# The Same Model, Written Twice

**Session 23 · no homework grades this · read it beside your own training loop**

Session 20 built a training loop by hand, and Session 21 varied it. By now the
loop is familiar enough to be boring, which is the moment a framework starts to
pay. This notebook writes one small model twice, in raw PyTorch and in
Lightning, on the same data with the same seed, and compares what each version
costs you in lines and what it buys. The `LightningModule` here has the same
methods as the one on the Session 23 slides (26-27).

**On Keras 3:** it is surveyed in the reading this term, not demonstrated and not
assessed. Maintaining three equivalent implementations of one idea costs more
than it teaches, so this notebook keeps two.

In [1]:
import logging
import time
import warnings

# Quiet console noise that says nothing about this model: the Trainer's banner
# lines, its advice to add DataLoader workers (on 1,347 rows of 64 features the
# loader is not the bottleneck), a progress-bar notice, and a deprecation
# inside Lightning itself. Real errors still show.
warnings.filterwarnings("ignore", message=".*does not have many workers.*")
warnings.filterwarnings("ignore", message=".*GPU available but not used.*")
warnings.filterwarnings("ignore", message=".*IProgress not found.*")
warnings.filterwarnings("ignore", message=".*LeafSpec.*")
logging.getLogger("lightning.pytorch").setLevel(logging.WARNING)

import numpy as np
import torch
import torch.nn.functional as F
import lightning as L
from torch import nn
from torch.utils.data import DataLoader, TensorDataset
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

print("torch", torch.__version__, "| lightning", L.__version__)
logging.getLogger("lightning.pytorch").setLevel(logging.WARNING)

digits = load_digits()
X = StandardScaler().fit_transform(digits.data).astype("float32")
y = digits.target.astype("int64")
X_train, X_valid, y_train, y_valid = train_test_split(
    X, y, test_size=0.25, random_state=0, stratify=y)

train_ds = TensorDataset(torch.from_numpy(X_train), torch.from_numpy(y_train))
valid_ds = TensorDataset(torch.from_numpy(X_valid), torch.from_numpy(y_valid))

def loaders(batch_size=64, seed=0):
    gen = torch.Generator().manual_seed(seed)
    return (DataLoader(train_ds, batch_size=batch_size, shuffle=True, generator=gen),
            DataLoader(valid_ds, batch_size=256))

def make_net():
    return nn.Sequential(nn.Linear(64, 64), nn.ReLU(), nn.Linear(64, 10))

print("train", X_train.shape, " validation", X_valid.shape)

torch 2.13.0 | lightning 2.6.5
train (1347, 64)  validation (450, 64)


## 1. Raw PyTorch

Every line here is one you have written before. Count what is about *your
problem*: the network (`make_net()`), the loss, the optimizer and its rate, and
the accuracy you score it by. Everything else, `zero_grad()`, `backward()` and
`step()` included, is protocol: the same in every project. That is the split
the slides call science and engineering.

In [2]:
EPOCHS = 12

torch.manual_seed(0)
model = make_net()
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-3)
loss_fn = nn.CrossEntropyLoss()
train_loader, valid_loader = loaders(seed=0)

t0 = time.perf_counter()
raw_history = []
for epoch in range(EPOCHS):
    model.train()
    for batch_x, batch_y in train_loader:
        optimizer.zero_grad()
        loss = loss_fn(model(batch_x), batch_y)
        loss.backward()
        optimizer.step()

    model.eval()
    correct = total = 0
    with torch.no_grad():
        for batch_x, batch_y in valid_loader:
            correct += int((model(batch_x).argmax(1) == batch_y).sum())
            total += len(batch_y)
    raw_history.append(correct / total)

raw_seconds = time.perf_counter() - t0
print(f"raw PyTorch  final validation accuracy {raw_history[-1]:.4f}  "
      f"in {raw_seconds:.1f}s")

raw PyTorch  final validation accuracy 0.9378  in 0.1s


## 2. The same thing, in Lightning

The `LightningModule` keeps the model, the loss, the metric and the optimizer
together; the `Trainer` owns the loop. Note what disappears: `zero_grad`,
`backward`, `step`, `model.train()`, `model.eval()`, `torch.no_grad()`, and the
device handling that this CPU example did not even need yet.

`validation_step` logs `val_acc` once per batch; Lightning averages those over
the epoch, weighting each batch by its size, so the logged number is the
accuracy over all 450 validation digits. A checkpoint or early-stopping callback
can only watch a metric that some step logs (section 5).

In [3]:
class DigitClassifier(L.LightningModule):
    def __init__(self, lr=1e-3):
        super().__init__()
        self.save_hyperparameters()
        self.net = make_net()

    def forward(self, x):
        return self.net(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        loss = F.cross_entropy(self(x), y)
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        acc = (self(x).argmax(1) == y).float().mean()
        self.log("val_acc", acc, prog_bar=True)

    def configure_optimizers(self):
        return torch.optim.AdamW(self.parameters(), lr=self.hparams.lr)

In [4]:
L.seed_everything(0, workers=True)
lit_model = DigitClassifier()
trainer = L.Trainer(max_epochs=EPOCHS, accelerator="cpu", logger=False,
                    enable_checkpointing=False, enable_progress_bar=False,
                    enable_model_summary=False, deterministic=True)

train_loader, valid_loader = loaders(seed=0)
t0 = time.perf_counter()
trainer.fit(lit_model, train_loader, valid_loader)
lit_seconds = time.perf_counter() - t0

lit_accuracy = float(trainer.validate(lit_model, valid_loader, verbose=False)[0]["val_acc"])
print(f"\nLightning    final validation accuracy {lit_accuracy:.4f}  in {lit_seconds:.1f}s")

Seed set to 0



Lightning    final validation accuracy 0.9378  in 0.2s


## 3. Are they the same model?

In [5]:
print(f"{'implementation':<16} {'val accuracy':>13} {'seconds':>9} {'parameters':>12}")
print(f"{'raw PyTorch':<16} {raw_history[-1]:>13.4f} {raw_seconds:>9.1f} "
      f"{sum(p.numel() for p in model.parameters()):>12,}")
print(f"{'Lightning':<16} {lit_accuracy:>13.4f} {lit_seconds:>9.1f} "
      f"{sum(p.numel() for p in lit_model.parameters()):>12,}")
n_valid = len(valid_ds)
raw_right = round(raw_history[-1] * n_valid)
lit_right = round(lit_accuracy * n_valid)
print(f"\nvalidation digits right: raw {raw_right} of {n_valid}, "
      f"Lightning {lit_right} of {n_valid}")
print("Same architecture, same optimizer, same seed, same batch order. A difference")
print("would come from shuffling or float non-determinism, not from the framework.")

implementation    val accuracy   seconds   parameters
raw PyTorch             0.9378       0.1        4,810
Lightning               0.9378       0.2        4,810

validation digits right: raw 422 of 450, Lightning 422 of 450
Same architecture, same optimizer, same seed, same batch order. A difference
would come from shuffling or float non-determinism, not from the framework.


## 4. What Lightning actually removed

In [6]:
removed = [
    "optimizer.zero_grad()",
    "loss.backward()",
    "optimizer.step()",
    "model.train() / model.eval()",
    "torch.no_grad() around evaluation",
    "the epoch loop and the batch loop",
    "manual accuracy accumulation across batches",
    "device placement (.to(device)) for model and every batch",
]
for line in removed:
    print("  -", line)
print(f"\n{len(removed)} pieces of protocol, none of them about this model.")

  - optimizer.zero_grad()
  - loss.backward()
  - optimizer.step()
  - model.train() / model.eval()
  - torch.no_grad() around evaluation
  - the epoch loop and the batch loop
  - manual accuracy accumulation across batches
  - device placement (.to(device)) for model and every batch

8 pieces of protocol, none of them about this model.


Two of those matter more than the rest:

- **`model.eval()` and `no_grad()`** are the ones people forget, and forgetting
  them is silent (Session 21 measured the cost).
- **Device placement** is invisible here because this is CPU-only. On a GPU it
  is the line that breaks a script when someone moves it to a different machine,
  and Lightning removes it entirely.

What you give up is directness: when something goes wrong inside `trainer.fit`,
you are debugging someone else's loop. That is a real cost, and it is why this
course teaches the loop first.

## 5. The features that are the actual argument

The boilerplate is a small win. Checkpointing, early stopping and multi-device
training are the reason people adopt Lightning: declared as callbacks rather
than written.

In [7]:
from pathlib import Path

from lightning.pytorch.callbacks import EarlyStopping, ModelCheckpoint

L.seed_everything(0, workers=True)
lit2 = DigitClassifier()
checkpoint = ModelCheckpoint(monitor="val_acc", mode="max", save_top_k=1,
                             dirpath="lightning_demo", filename="best")
stopper = EarlyStopping(monitor="val_acc", mode="max", patience=4)

trainer2 = L.Trainer(max_epochs=60, accelerator="cpu", logger=False,
                     callbacks=[checkpoint, stopper], enable_progress_bar=False,
                     enable_model_summary=False, deterministic=True)
train_loader, valid_loader = loaders(seed=0)
trainer2.fit(lit2, train_loader, valid_loader)

print(f"\nasked for 60 epochs, stopped after {trainer2.current_epoch}")
print(f"best validation accuracy: {float(checkpoint.best_model_score):.4f}")
print(f"checkpoint written to    : {Path(checkpoint.best_model_path).name}")

Seed set to 0



asked for 60 epochs, stopped after 28
best validation accuracy: 0.9756
checkpoint written to    : best.ckpt


In [8]:
import shutil
shutil.rmtree("lightning_demo", ignore_errors=True)
print("demo checkpoint directory removed")

demo checkpoint directory removed


Two callback objects and one `callbacks=` argument: that is the whole cost of
behaviour that took a page of bookkeeping in Session 21's harness, and the same
lines work whether you are on one CPU or eight GPUs. Both callbacks watch
`val_acc`, which exists only because `validation_step` logs it.

## What to take from this

- Lightning does not change the model, the optimizer or the result. It removes
  the protocol around them.
- What it removes is exactly the part that is easy to get silently wrong:
  `eval()` mode, `no_grad()`, device placement.
- The cost is indirection when something breaks, which is why you learn the
  loop first and adopt the framework second.
- Checkpointing and early stopping as callbacks are the real argument, not the
  line count.
- Keras 3 is this term's reading-only survey: it is a third way to write the
  same thing, and it is not assessed.

## Where to go next

- **Reading, Session 23**: transfer learning, Lightning, and the Keras 3
  survey.
- **`02_pytorch_cnn.ipynb`, Part 3**: the ResNet transfer recipe from the
  Session 23 slides (pinned weights, freeze, replace the head), run on MNIST.
- **Your project**: if your team starts from a pretrained model, the same
  transfer recipe applies there, head first and then deeper layers; the project
  guide says what is graded.